### ISSUE TITLE
[Feature Request] Benchmark mode

### ISSUE THREAD

**Issue by parkan** [filecoin-project/boost#498]
One of the biggest issues I have currently with selecting providers to work with is determining throughput between my node and their miners, both before adding them to my provider list in Estuary and during ongoing transfers. We see highly variable speeds to different nodes (10-20x difference between superficially similar miners at times) so this is important for us, especially at 100TiB+ dataset sizes.

To the best of my knowledge, the tools available to screen a miner before adding them are pretty limited:
- ping/traceroute: very limited, does not directly represent throughput, can sometimes be helpful if you know that particular peerings have issues
- find a ookla/speedtest server in the same AS as the miner, run `speedtest-cli` against it: total hack, limited coverage, still not that accurate
- iperf3: best available option but requires active coordination with miner's ops team, potentially leaves them open to portscans/bandwith wasting
- making some online deals and doing the math: not great
- various miner stats dashboards: no help here as they don't reflect the network environment between the client and miner

Boost could be an interesting place to house a native in-protocol benchmark tool to quickly evaluate a set of miners for inclusion in your active set, as it's a standalone tool (will not bloat the main executable, portable) and a logical place to house utility functionality related to transfers. The tests already appear to contain a dummy http server, so that can be used to serve the bench payload or else a test file/url could be specified.

The command could look as simple as `boost -vv bench --provider=<f00001>` if the boost machine and the intended HTTP host are the same (as as likely the case with Estuary), outputting standard iperf style stats. A tiny amount of FIL can be spent to avoid making this a DOS vector.

This would also help https://github.com/filecoin-project/boost/issues/208#issue-1120649180 (my dream benchmark tool would parametrize all of these)

Does this make sense within the scope of Boost? There is of course the option of just asking the miner to curl a file from your HTTP endpoint and post results, but that's only marginally less onerous than iperfing.

**Comment by dirkmc** [comment]
I think it would be hard to motivate storage providers to participate, and if they did participate they may give different QoS to these particular requests. It also wouldn't account for their sealing throughput, which is likely to be the true bottleneck.

I believe the primary estuary deployment keeps track of statistics for the storage providers that it makes deals with, and adjusts who it sends deals to accordingly. Something like this would probably be a more reliable strategy.

**Comment by parkan** [comment]
> I think it would be hard to motivate storage providers to participate, and if they did participate they may give different QoS to these particular requests

That's fair and part of why I suggested a small FIL payment to encourage participation, though this is probably a whole can of worms as it gets into the economics of the system.

> It also wouldn't account for their sealing throughput, which is likely to be the true bottleneck.

I think a lot of the performance work has been based on this assumption (sealing as bottleneck), but it has not been my experience. My guess is that PL-operated clients are generally positioned on big machines with top tier connectivity, which does result in ready saturation of sealing capacity on the miner side. In my case, I'm working with more limited hardware and funky peering, which has resulted in a host of client-side bottlenecks well before we hit major miner sealing limits -- compute for car packing/commP, disk, etc. I've worked through compute/memory/IO limitations as those are contained on our end, but network transfers left me stumped as the number of variables is much higher. I've seen everything from bad on-chain peer data, to routing failures, to stream resets, highly variable transfer rates, etc.

So at this point I'm really just trying to get my hands on diagnostic tools for the network segment.

And of course the various miner perf dashboards already show reasonably useful sealing capacity data :slightly_smiling_face: 

> I believe the primary estuary deployment keeps track of statistics for the storage providers that it makes deals with, and adjusts who it sends deals to accordingly. Something like this would probably be a more reliable strategy.

I'm currently basing my providers on the active set from estuary.tech, but even with that pre-filtering lots of them just don't work out and I have to do a ton of babysitting. It does look like some tooling may have been stood up since my last sync with the estuary team, so I'm going to check in with them and see what's new.
